### ISSUE TITLE
Correlations for fixed effects being flagged

### ISSUE THREAD

**Issue by mattfidler** [nlmixrdevelopment/nlmixr#243]
Hi @billdenney ,

After our discussion at ACoP, I am adding an issue that correlations should flagged like [sumo](https://github.com/UUPharmacometrics/PsN/releases/download/4.9.0/sumo_userguide.pdf) in PsN.

Using NONMEM jargon, this is already done for `omega` matrices but not for correlations in `theta` matrices.  Currently `nlmixr` points out moderate correlations (|r| > 0.3) and strong correlations (|r| > 0.7), though the numbers are currently hard-coded (this probably should be changed)

Also the ability to do this should be present in nlmixr for `theta` estimates as measured by their correlation matrix.

**Comment by mattfidler** [comment]
Fix typo in 600c08f024409f0b7f5b41e9836bb7cbb15287e3

**Comment by billdenney** [comment]
Here is some sumo output for comparison:

```
Successful minimization                                           [    OK   ]
No rounding errors                                                [    OK   ]
No zero gradients                                                 [    OK   ]
No final zero gradients                                           [    OK   ]
Hessian not reset                                                 [    OK   ]
No parameter near boundary                                        [    OK   ]
Covariance step                                                   [    OK   ]

Condition number                                                  [    OK   ]
Correlations                                                      [    OK   ]
```

**Comment by billdenney** [comment]
Are your `omega` correlations the correlations at the level of the sampled random effects (i.e. if there should be a correlation within the matrix estimate) or a correlation of the estimated magnitude of the estimated standard deviation?

**Comment by mattfidler** [comment]
They are at the level of the random effects, the estimated values.  Standard errors are not calculated on the variance parameters since they do not follow a multivariate normal.

Note the following do not exactly make sense to `nlmixr`:

- Rounding Errors;  While we could calculate "significant digits" the same way NONMEM does, they do not play into the minimization.

The following warnings are already output from `FOCEi`
-  `Hessian reset during optimization; (Can control by foceiControl(resetHessianAndEta=.))`
- `Initial ETAs were nudged; (Can control by foceiControl(etaNudge=.))`
- `ETAs were reset to zero during optimization; (Can control by foceiControl(resetEtaP=.))`
- `Tolerances were reduced during Gill Gradient, so it was repeated %d/%d times\nYou can control this with foceiControl(repeatGillMax=.)`
-  `Gradient problems with initial estimate and covariance; see $scaleInfo`
- `Gradient problems with initial estimate; see $scaleInfo`
- `Gradient problems with covariance; see $scaleInfo`
-  `Tolerances (atol/rtol) were reduced (after %d bad solves) for some difficult ODE solving during the optimization.\nCan control with foceiControl(stickyRecalcN=)\nConsider reducing sigdig/atol/rtol changing initial estimates or changing the structural model.`
- `Tolerances (atol/rtol) were temporarily reduced for some difficult ODE solving during the optimization.\nConsider reducing sigdig/atol/rtol changing initial estimates or changing the structural model.`
- `mu-referenced Thetas were reset during optimization; (Can control by foceiControl(resetThetaP=.,resetThetaCheckPer=.,resetThetaFinalP=.))`
- `Non-positive definite individual Hessian at solution(ID=%d); FOCEi objective functions may not be comparable.`  This is because the Hessian calculation uses generalized cholesky inverses
- `Bad solve during optimization.`
- `R matrix non-positive definite`
- `Hessian (R) matrix seems singular; Using pseudo-inverse`
- `S matrix seems singular; Using pseudo-inverse`
- `R matrix non-positive definite but corrected by R = sqrtm(R%*%R)`
- `R matrix non-positive definite but corrected (because of cholAccept)`
- `S matrix non-positive definite but corrected by S = sqrtm(S%*%S)`
-  `S matrix non-positive definite but corrected (because of cholAccept)`
- `Since sandwich matrix is corrected, you may compare to $covR or $covS if you wish.`
- `Using R matrix to calculate covariance, can check sandwich or S matrix with $covRS and $covS`
- `Parameter estimate near boundary; covariance not caculated. Use getVarCov to calculate anyway.`

Missing warnings for FOCEi:
- `No zero gradients` (only applicable in gradient methods, can use `bobyqa` where it doesn't matter)
- `No final zero gradients`
- `Cannot calculate covariance.` (It is currently a printout instead).

**Comment by mattfidler** [comment]
Now it prints out high correlated parameters ">0.7", though 0.9 may be a more standard limit for pharmacometrics.  I'm unsure where I got the numbers now.

**Comment by mattfidler** [comment]
I can find references that back up the numbers I use, though I can find references for others.

**Comment by billdenney** [comment]
Thanks for this detail.  The one question that remains open to me is on the correlation of estimated `omega` magnitude.  While the parameter follows a Wishart distribution, if the estimate is highly correlated with other parameters, it can cause minimization trouble.  (This is simple to investigate in a Bayesian sense via correlation examined visually with a pairs scatter plot or equivalent calculations therein.)

That is the type of warning that I’m hoping for.  Also, when simulating VPCs or similar with parameter uncertainty, I’d like to include that correlation.

The explanation you gave above makes sense at the printout level, but I think that I may be missing something for the diagnostic level.

**Comment by mattfidler** [comment]
I agree that one reason for problems in minimization is highly correlated parameters.  However, it is also difficult to get a successful covariance step with highly correlated parameters.  So I'm unsure standard errors on `omega` parameters can help diagnose this problem.  

Looking at the `eta` parameter correlations should give the same information and is not dependent on a successful covariance step but it is confounded by shrinkage.  

I am not sure the best metric here.  

On a related issue, the current calculation of condition number excludes omega standard error values as well.  However, the condition number is also calculated on the unscaled parameter space, which is also not really a good reflection of the condition of the system during minimization.  The scaling greatly affects the curvature of the problem (hence scaling is done to help minimize the system). In my view the condition number calculated in the unscaled space is not accurate reflection of the system.  This is my insight working with `focei`, though it is un-tested.

**Comment by billdenney** [comment]
You’ve given a lot to think about here.

The last is the easiest for me to think about.  While the condition number is in parameter and not scaled space, I’d assume that the differences in curvature would have the biggest impact in parameter search and not asymptotic performance at the final estimate.  (In other words, I would be surprised if it were different in most cases.)

Not including eta values there seems like it could cause issues, but I typically find that etas affect fixed effects to a degree that the model would be unstable in a fixed effect most of the time if the instability were coming from an eta.  (But, I’m sure that someone can find a counter-example.)

**Comment by mattfidler** [comment]
I forgot to add my thoughts about `VPC`s

>Also, when simulating VPCs or similar with parameter uncertainty, I’d like to include that correlation.

Using an inverse Wishart simulation takes into consideration the uncertainty in the whole `omega`, including the uncertainty in the correlations. 

Here is an example of the uncertainty simulation used in `RxODE` for `50` and `1000` subjects:

<details>

``` r
library(RxODE)
set.seed(100)
cvPost(50,diag(3), 3)
#> [[1]]
#>             [,1]       [,2]        [,3]
#> [1,]  1.16269834 -0.2042712 -0.06369004
#> [2,] -0.20427119  1.4932645  0.10406012
#> [3,] -0.06369004  0.1040601  1.03861747
#> 
#> [[2]]
#>             [,1]       [,2]        [,3]
#> [1,]  0.90247461 -0.0963300 -0.09452588
#> [2,] -0.09633000  1.2633626  0.14227213
#> [3,] -0.09452588  0.1422721  0.95161119
#> 
#> [[3]]
#>            [,1]       [,2]        [,3]
#> [1,] 0.82623204  0.0247928  0.09502786
#> [2,] 0.02479280  1.3001922 -0.34062132
#> [3,] 0.09502786 -0.3406213  0.95999836
cvPost(1000,diag(3), 3)
#> [[1]]
#>            [,1]       [,2]       [,3]
#> [1,] 1.02116256 0.00107742 0.01534162
#> [2,] 0.00107742 0.96163859 0.02496015
#> [3,] 0.01534162 0.02496015 0.95134864
#> 
#> [[2]]
#>             [,1]       [,2]        [,3]
#> [1,]  1.02923782 0.02697585 -0.01417398
#> [2,]  0.02697585 1.05616392  0.01167958
#> [3,] -0.01417398 0.01167958  0.98626284
#> 
#> [[3]]
#>             [,1]        [,2]        [,3]
#> [1,]  1.01131226 -0.06015239 -0.02612809
#> [2,] -0.06015239  1.01380166  0.04188760
#> [3,] -0.02612809  0.04188760  0.98888150
```

<sup>Created on 2019-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup> 

</details>

Note the covariance has correlations even though the estimated matrix does not have correlations.  As the certainty goes up, the simulated omega matrix parameters more closely approach the estimated matrix (in this case the identity matrix)

Note that the correlations are based on the original estimated matrix.

Since this doesn't make the assumption the `omgea` matrix is a truncated normal distribution (which is incorrect) AND it also will always be positive definite, I think this is the correct way to simulate uncertainty in the `omega` covariance matrix.  It has the added benefit of being based on statistical (Bayesian) theory.

On the other hand, using the correlation in omega parameters doesn't exactly make sense to me.  To me, if there is a sizable correlation, then you should include it in the `omega` matrix itself.

Other than theoretical considerations, I am not sure if it is going to give drastically different results by simulating under the wrong assumptions.  I believe `monolix` uses this assumption (though their documentation on this is sparse) and many modelers using NONMEM use this assumption.  Still I think simulating from the proper distribution makes the most sense.  If you have any articles that show something else I would like to see them.

One last note, simulating `VPC`s with parameter uncertainty turns them into something else.  I don't know the correct term for it, but it isn't a `VPC`.

**Comment by mattfidler** [comment]
The example in the tutorial also shows how to simulate from a nlmixr model with uncertainty;  See:

https://nlmixrdevelopment.github.io/nlmixr/articles/xgxr-nlmixr-ggpmx.html#simulation-of-a-new-scenario-with-rxode

**Comment by mattfidler** [comment]
> The last is the easiest for me to think about. While the condition number is in parameter and not scaled space, I’d assume that the differences in curvature would have the biggest impact in parameter search and not asymptotic performance at the final estimate. (In other words, I would be surprised if it were different in most cases.)

I think you are likely correct here.

> Not including eta values there seems like it could cause issues, but I typically find that etas affect fixed effects to a degree that the model would be unstable in a fixed effect most of the time if the instability were coming from an eta. (But, I’m sure that someone can find a counter-example.)

I don't follow this.  Sorry.

**Comment by billdenney** [comment]
I wasn’t clear in the part that you didn’t follow.  I’m not using terminology well for separating inter-individual variability as an estimate of the standard deviation (or equivalently of the variance) and as the specific post-hoc instance for an individual.  I’m going to call them the sd(eta) (estimate of standard deviation) and etai (individual post-hoc) to be clearer here.  (Also, I’m writing this on my phone; it wants to autocorrect a lot of the terminology here.  If something looks weird, please ask as it may be a phone-induced typo.)

I think of sd(eta) and etai in completely separate ways.

If etai is correlated between parameters, I need to estimate the correlation to achieve a good estimate of sd(eta).  For that, I agree with the limits you have suggested where correlation of etai between different parameters needs a correlation term.

For the correlation of sd(eta) either with other sd(eta) values or with fixed or residual effects, that is a separate problem.  When sd(eta) is correlated with another parameter, it can break minimization such that it fails as with almost any other parameter being correlated.  In my experience, this can happen when I try to fit too many sd(eta) on too many parameters or when my fixed effects start to take away all of the variability on a parameter.  I had the latter happen when I was working on a study, and the fixed effects plus residuals explained all the data and the random effects started to try to become residual error (or that is what it seemed like, anyway).

Clearer?  (Hopefully?)

**Comment by billdenney** [comment]
I think that a VPC with parameter uncertainty becomes a prediction interval.

**Comment by mattfidler** [comment]
>Clearer? (Hopefully?)

Yes.  Currently the methods we use do not provide `sd(eta)` correlations with any other parameters. 

How do you know if it is "breaking" the minimization if you cannot minimize?  Usually a covariance step is provided at the end of the minimization.  If we employ your method here, then perhaps there may be a way to add this to the check?

> I think that a VPC with parameter uncertainty becomes a prediction interval.

Perhaps,  it seems to fulfill the requirements;  At the same time I am not exactly sure I have a rigorous enough proof that it is a prediction interval to call it that...

**Comment by billdenney** [comment]
"Breaking" was not a great word for what I was trying to describe for what parameters which impede convergence are doing.  I think of parameters that cause convergence to take significantly longer due to a more difficult search with flattening of much of the parameter space as potentially "breaking" the minimization.  This is because the parameter space flattening (often via the introduction of correlation between many parameters) can cause minimization to give inaccurate results in this flattened space.

One symptom of this flattening is an issue with correlation between many parameters.  Often, the correlation is with fixed effects, but sometimes the correlation is with `sd(eta)` or the residual error.  In one current example for me, including additive error in the model makes it more accurate in that a 3-compartment model can be fit, but the asymptotic covariance matrix explodes.  The condition number goes from about 20 to 1e6, but in almost any other metric, the model is better.

I fear that I'm beating a dead horse here (intended as a negative for me, not you), but where I'm going with all of this is that I'd like to be able to both help diagnose where an issue is coming from (as the "breaking" discussion above is going) and to sample asymptotic uncertainty on all parameters (as the VPC/prediction interval discussion is going).

**Comment by mattfidler** [comment]
Not all algorithms will ever support `SEs` on `sd(eta)`, for instance `nlme`.

**Comment by mattfidler** [comment]
I may consider it in the future
